### PR TITLE
Extended dshot telemetry data parsing and presentation

### PR DESCRIPTION
**Pull request by damosvil**
This pr adds EDT parsing and presentation functionality to blackbox explorer.
EDT data is packed in debug fields in the following way:
- bit 15 - alert event
- bit 14 - warning event
- bit 13 - error event
- bits 11-8 - Max stress level during the whole flight
- bits 7-0 - ESC data depending on the chosen debug mode

This pr is related to:
https://github.com/bird-sanctuary/bluejay/pull/64
https://github.com/betaflight/betaflight/pull/12170
https://github.com/betaflight/betaflight-configurator/pull/3262

ESC temperature log files for testing:
[BTFL_BLACKBOX_LOG_Meteor65_20230201_183226.zip](https://github.com/betaflight/blackbox-log-viewer/files/10572044/BTFL_BLACKBOX_LOG_Meteor65_20230201_183226.zip)
[BTFL_BLACKBOX_LOG_Meteor65_20230202_183845.zip](https://github.com/betaflight/blackbox-log-viewer/files/10572056/BTFL_BLACKBOX_LOG_Meteor65_20230202_183845.zip)

ESC demag metric files for testing:
[BTFL_BLACKBOX_LOG_Meteor65_20230203_224822.zip](https://github.com/betaflight/blackbox-log-viewer/files/10583249/BTFL_BLACKBOX_LOG_Meteor65_20230203_224822.zip)

ESC rpm log files for testing:
[BTFL_BLACKBOX_LOG_Meteor65_20230203_232019.zip](https://github.com/betaflight/blackbox-log-viewer/files/10583252/BTFL_BLACKBOX_LOG_Meteor65_20230203_232019.zip)

### CODE REVIEW

**Comment by nerdCopter** [comment]
* unnecessary spaces added, can you set your editor to trim trailing spaces on new code, or edited lines only?
image

* can you set your editor to convert tabs to spaces, preferably to match existing code? (tab = 4 spaces)
image

**Comment by damosvil** [comment]
Replaced tabs by spaces in modified files

**Comment by McGiverGim** [comment]
@damosvil I don't understand too mucho about ESCs but an important question to get to the best solution:
- This info about demag, desync, stall, etc. events. How important is? Is habitual to have them and not notice it or if it happens is sign that something is broken/misconfigured and needs to be fixed?

I say that because package it only in debug fields, will "hide" it from the major part of users, except if they configure this debug fields in the firmware and look at it expressly in the blackbox.
- If this events are habitual, at least at some extent -> then use debug fields is ok to me.
- If this events are not, and must be noticed by the user -> then I think is better to send them as events and not as debug fields, in this way they will be noticed inmediatelly in the blackbox (but we need to revisit the events fields because they are limited to 32 only I think remember). The debug fields may remain, they can give more information if a user needs it, but must be not the principal way of notification.

Regards!

**Comment by McGiverGim** [comment]
@damosvil looking deeply at the code, it seems you are using the bookmarks feature to mark them? Maybe you can publish some screen capture about how they look?

**Comment by damosvil** [comment]
Below I describe behavior of  some of the events:
- Demag events are pretty usual as they happen when motors change speed abruptly. 
- Desync events are triggered when many consecutive demag events happen (in Bluejay/Blheli_S):
-- 160 consecutive demag events for Low Demag Compensation trigger a Desync event
-- 130 consecutive demag events for High Demag Compensation trigger a Desync event
-- 255 consecutive demag events for Off Demag Compensation trigger a Desync event
- Stall events happen when motor stalls
- Demag Metric data is used to adjust commutation timing. It ranges from 120 to 255. When it is low it means that the ESC is capable of doing good commutation, the higher the worse commutation.
- Max Demag Metric is a value between 0 and 15 that indicates how the esc carries on with the motor during flight. It is the result of the formula: MDM = (DM - 120) / 9. During the test flights in a brand new Meteor65 it had a value of 4. Normal values should be below 10, otherwise the ESC may have problems performing a good commutation.

Debugging EDT you will always have the following status data:
- Demag event
- Desync event
- Stall event
- Max Demag Metric

You will have an additional value depending on the debug mode used:
- DSHOT_STATUS_N_TEMPERATURE: Status data and temperature
- DSHOT_STATUS_N_VOLTAGE: Status data and temperature
- DSHOT_STATUS_N_CURRENT: Status data and current
- DSHOT_STATUS_N_DEBUG1: Status data and debug1 value
- DSHOT_STATUS_N_DEBUG2: Status data and debug2 value
- DSHOT_STATUS_N_DEMAG_METRIC: Status data and demag metric
- DSHOT_STATUS_N_ERPM_FRACTION_18: Status data and RPM but with less resolution than when debugging DSHOT_RPM_TELEMETRY

This is the first time all this data is logged in Betaflight blackbox, so any idea regarding on how to display it is more than welcome.

Capture of temperature:
imagen

Capture of demag metric on a brand new Meteor 65:
imagen

**Comment by damosvil** [comment]
At this time I choose to draw the EDT chosen magnitude in the graph, but I am thinking on print a mark in the temperature graph for desync events and a line from top to bottom for stall events. Demag events are pretty usual so I would let them in the info to the right, but I would not print them. Max demag metric is interesting to be checked at the end of the flight, so I would not print this number in the graph, it can be simply cheched at the end of the flight.
Please, let me know how you see this.

**Comment by damosvil** [comment]
In general all this data is interesting to technical users that want to know how the esc/motors are going during flight.
The most interesting to be tracked during the flight is max demag metric. It can provide information about the esc/motor health during the whole flight.

**Comment by McGiverGim** [comment]
Ok, understood. So the major part of values from the ESC are debug values and must be shown when the users logs it. As you say this PR is ok for that.

But it seems there are some information that can be important to show (maybe desync/stall events? can you confirm?) It seems to me, by your description, that they must not happen and if it does, is a good idea to inform the user. Am I right?

If yes, I think they must be shown as events in the timeline. Is similar to what we do with the flight modes, that we draw a "flag/mark" in the timeline: 
https://github.com/betaflight/betaflight/blob/facf44b4066d726fafcc969a8fcb9afbeb3009ff/src/main/blackbox/blackbox.c#L354
but this can be left to a future PR. Adding the debug data as first step is ok to me.

**Comment by McGiverGim** [comment]
Looking at the code, the `lapTimer` must be taken as a good example too. But as I said, can be left for a future PR.

**Comment by haslinghuis** [comment]
@damosvil please remove dependency changes as they are not required for your (excellent) work.

**Comment by damosvil** [comment]
@haslinghuis I removed the changes over package.json and yarn.lock. Please, let me know if I have to change anything else.

File: js/flightlog.js
Comment (McGiverGim) [inline js/flightlog.js:311]: I'm not too sure about this. I think in the rest of the code we simply store the latest value, and add them when we process other kind of data, to add the values while processing. Maybe is clear with your method, with a little more process time (it needs to look at all the data chunk). 

Can you look at the slow frames and gps frames chunks to see if they can be "filled" with this and remove the old implementation? Taking times of this method will be great to see if we loose too much time. If it takes not too much time, I prefer your implementation.

When there are errors in the blackbox (frames lost), this method will fill the data? I think remember that not, that we don't have data in the chunk when the frames are lost.

**Review by McGiverGim — DISMISSED** [review]

File: js/flightlog_fields_presenter.js
Comment (damosvil) [reply js/flightlog_fields_presenter.js]: Done

File: js/flightlog_parser.js
Comment (haslinghuis) [inline js/flightlog_parser.js:1120]:
```suggestion
    function assimilateDebugMode(sysConfig, frameDef) {
        
        console.log("Debug mode is " + DEBUG_MODE[sysConfig.debug_mode] + ":" + sysConfig.debug_mode);
        
        if (DEBUG_MODE[sysConfig.debug_mode].startsWith("DSHOT_STATUS_N_")) {
            for (let k = 0; k < frameDef.name.length; k++) {
                if (frameDef.name[k].startsWith("debug")) {
                    // Assimilate encoding depending on debug mode
                    frameDef.encoding[k] = FLIGHT_LOG_FIELD_ENCODING_PACK_1F_1F_1F_1G_4U_8U;
                    
                    // Assimilate predictor depending on debug mode
                    switch (DEBUG_MODE[sysConfig.debug_mode]) {
                    case "DSHOT_STATUS_N_VOLTAGE":
                        frameDef.predictor[k] = FLIGHT_LOG_FIELD_PREDICTOR_DSHOT_STATUS_N_VOLTAGE;
                        break;
                    case "DSHOT_STATUS_N_ERPM_FRACTION_18":
                        frameDef.predictor[k] = FLIGHT_LOG_FIELD_PREDICTOR_DSHOT_STATUS_N_ERPM_FRACTION_18;
                        break;
                    default:
                        break;
                    }
                }
            }
        }
    }
```

Comment (damosvil) [reply js/flightlog_parser.js]: Done

**Comment by damosvil** [comment]
Please, consider this log to better view the latest functionality changes on graph:
[BTFL_BLACKBOX_LOG_Meteor65_20230208_183723.zip](https://github.com/betaflight/blackbox-log-viewer/files/10776397/BTFL_BLACKBOX_LOG_Meteor65_20230208_183723.zip)

File: js/flightlog.js
Comment (damosvil) [reply js/flightlog.js:311]: > Can you look at the slow frames and gps frames chunks to see if they can be "filled" with this and remove the old implementation? Taking times of this method will be great to see if we loose too much time. If it takes not too much time, I prefer your implementation.

I have been looking for the current implementation. Is it this?
imagen


> When there are errors in the blackbox (frames lost), this method will fill the data? I think remember that not, that we don't have data in the chunk when the frames are lost.
Still looking into this

**Comment by nerdCopter** [comment]
@damosvil 
image

**Comment by damosvil** [comment]
@nerdCopter I have seen that bug, but the program is using that code path. I don't understand why it is reporting it. I will have a look asap.

**Comment by nerdCopter** [comment]
seemingly:
Untitled

**Comment by damosvil** [comment]
@nerdCopter the continues that you mention aren't related to the external loop. I think they are outside the loops you are refering to, I mean that if they were related to these loops shouldn't they be written like in the following example?
```
for (;;) {
  command = command + 1;
  continue;
}
```

**Comment by nerdCopter** [comment]
> @nerdCopter the continues that you mention [...]

i think you're right and i'm wrong; i'm investigating.  it may be SonarCloud false-positive.

**Comment by nerdCopter** [comment]
research reports that `continue` is only for for loops.  i see one at line 1211 without a for loop. i wonder if this is what breaks SonarCloud. 
i tried setting up sonarclod on my own repo, but it's convoluted and may require further CI configs which i dont feel like performing.

**Comment by haslinghuis** [comment]
Status update pls

**Comment by damosvil** [comment]
This weekend I will update the code to avoid sonarcloud false positives, and allow being integrable in current master

**Comment by damosvil** [comment]
Today I generated a log in a Meteor 65 and it seems the pr works fine:
[STATUS_N_RPM_BTFL_BLACKBOX_LOG_Meteor65_20230318_104401.zip](https://github.com/betaflight/blackbox-log-viewer/files/11008198/STATUS_N_RPM_BTFL_BLACKBOX_LOG_Meteor65_20230318_104401.zip)

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Comment by damosvil** [comment]
@haslinghuis @nerdCopter 

Hi guys,

Please, would it be possible to generate a installation bundle?
We at Bluejay would like to give people tools to test new version, and we need they submit and visualize the logs.

**Comment by nerdCopter** [comment]
i produced similar results with these settings using your data:
image


lines look debug traces look slightly different for me (thicker in some places) (not a problem i presume)
examples aligned with your previous screens
image

i find these artifacts odd, but maybe it's specific to my machine 🤷‍♂️
image

i skimmed over code not diving too deeply, but overall the code seems fine.

your editor auto-fixed lines of code that were outside of this PR however.  some editors can't be micromanaged, but some can be set to only auto-correct new line and modified lines.

**Comment by damosvil** [comment]
@nerdCopter those artifacts are intentional to show when alert (wide line) and warning (inverted triangle) esc events happen.
I still have to make all the corresponding documentation. I will document all of this when Bluejay 0.20 is finished

**Comment by damosvil** [comment]
To produce data please check:
https://github.com/bird-sanctuary/bluejay/wiki/0.20-Testing-Guide-and-Checklist

**Comment by nerdCopter** [comment]
> To produce data please check:
> [bird-sanctuary/bluejay/wiki/0.20-Testing-Guide-and-Checklist](https://github.com/bird-sanctuary/bluejay/wiki/0.20-Testing-Guide-and-Checklist)

my 16-bit does not have blackbox; i'll have to leave extended testing to others.

**Comment by damosvil** [comment]
Here are some logs from a user called Zoggbarr:
[202303251.zip](https://github.com/betaflight/blackbox-log-viewer/files/11070275/202303251.zip)

**Comment by nerdCopter** [comment]
* seemingly does everything it's supposed to.
* recommended don't merge until https://github.com/betaflight/betaflight/pull/12170 and https://github.com/betaflight/betaflight-configurator/pull/3262 are merged.

**Comment by McGiverGim** [comment]
Added the `pinned` tag to not let the bot close this...

File: js/flightlog.js
Comment (haslinghuis) [inline js/flightlog.js:326]: Please try to avoid using var in new code. Braces are a part of the thought containing the conditional or loop. Therefore, they belong on the same line.
```suggestion
        let previousFrame = 0, currentFrame;
        
        for (let k = 0; k < chunks.length; k++) {
            for let (j = 0; j < chunks[k].frames.length; j++) {
                currentFrame = chunks[k].frames[j]; 
                if (previousFrame === 0)
                    previousFrame = currentFrame;
                    
                for (let i = 0; i < currentFrame.length; i++) {
                    if (!Array.isArray(currentFrame[i]) && Array.isArray(previousFrame[i])) {
                        currentFrame[i] = previousFrame[i];
                    }
                }
                
                previousFrame = chunks[k].frames[j];
            }
        }
    }
```

File: js/flightlog_fielddefs.js
Comment (haslinghuis) [inline js/flightlog_fielddefs.js:356]: Missing debug modes should be added in separate PR as probably needs to be added using `adjustFieldDefsList` function

File: js/flightlog_fields_presenter.js
Comment (haslinghuis) [inline js/flightlog_fields_presenter.js:1032]: Copy pasta error?

File: js/flightlog_fields_presenter.js
Comment (haslinghuis) [inline js/flightlog_fields_presenter.js:1060]: Let's add missing items in separate PR to keep intent of current PR isolated.

File: js/flightlog_fields_presenter.js
Comment (haslinghuis) [inline js/flightlog_fields_presenter.js:1899]: As above

File: js/flightlog_parser.js
Comment (haslinghuis) [inline js/flightlog_parser.js:1144]: Try not adding new vars - please declare / initialize at block scope.

File: js/graph_config.js
Comment (haslinghuis) [inline js/graph_config.js:1021]: please remove

Comment (damosvil) [reply js/graph_config.js]: done

File: js/flightlog_parser.js
Comment (damosvil) [reply js/flightlog_parser.js:1156]: The function where this i, j variables are defined use them in loops and after the loops, so in this case I recommend keeping them. They are defined that way because of the logic in this function.

File: js/flightlog_fielddefs.js
Comment (damosvil) [reply js/flightlog_fielddefs.js]: resolved in current codebase

File: js/flightlog_fields_presenter.js
Comment (damosvil) [reply js/flightlog_fields_presenter.js]: resolved in current codebase

File: js/flightlog_fields_presenter.js
Comment (damosvil) [reply js/flightlog_fields_presenter.js]: resolved in current codebase

File: js/flightlog_fields_presenter.js
Comment (damosvil) [reply js/flightlog_fields_presenter.js]: resolved in current codebase

File: js/flightlog.js
Comment (damosvil) [reply js/flightlog.js]: refactored

File: js/flightlog.js
Comment (haslinghuis) [reply js/flightlog.js:304]: Has this been resolved?

Comment (haslinghuis) [reply js/flightlog.js:304]: @McGiverGim are you okay with this - for a follow up PR?

**Review by McGiverGim — APPROVED** [review]

**Comment by nerdCopter** [comment]
please merge master or rebase master if this is still a desirable PR.

**Comment by stylesuxx** [comment]
@nerdCopter is there any way I can take over this PR? I don't think Daniel will be able to work on this any time soon.

I could fork his changes and submit a new PR, unless you have a better idea?

**Comment by nerdCopter** [comment]
@stylesuxx , give me a short stint of time, i can update it and push. if it is not up to par and needs further coding, then i can hand it off.

**Comment by nerdCopter** [comment]
@stylesuxx , okay, i can easily update the branches but it turns out i cannot push to the existing PR's.

If you would like to take over all 3 PR's, please be our guest.  Do note that rather than MSP semver `1.46` comments/code, it will need to be `1.47`.

The idea here is to checkout these branches and either merge-master or rebase master (your choice) then make _new_ PR's from them.  Although i'm certainly able to do the same, i presume there may need further technical changes from reviews that i will not be competent to accomplish.

TYVM.

**Comment by haslinghuis** [comment]
This work will be continued in a new PR.

**Pull request closed without merging**
